air/src/utils.rs

```diff
@@ -0,0 +1,224 @@
+//! A collection of utility functions helpful in defining AIR's.
+
+use core::array;
+
+use p3_field::{AbstractField, Field};
+
+use crate::AirBuilder;
+
+/// Pack a collection of bits into a number.
+///
+/// Given vec = [v0, v1, ..., v_n] returns v0 + 2v_1 + ... + 2^n v_n
+#[inline]
+pub fn pack_bits_le<AF, Var, I>(iter: I) -> AF
+where
+    AF: AbstractField,
+    Var: Into<AF> + Clone,
+    I: DoubleEndedIterator<Item = Var>,
+{
+    let mut output = AF::ZERO;
+    for elem in iter.rev() {
+        output = output.double();
+        output += elem.clone().into();
+    }
+    output
+}
+
+/// Computes the arithmetic generalization of boolean `xor`.
+///
+/// For boolean inputs, `x ^ y = x + y - 2 xy`.
+#[inline(always)]
+pub fn xor<AF: AbstractField>(x: AF, y: AF) -> AF {
+    x.clone() + y.clone() - x * y.double()
+}
+
+/// Computes the arithmetic generalization of a triple `xor`.
+///
+/// For boolean inputs `x ^ y ^ z = x + y + z - 2(xy + xz + yz) + 4xyz`.
+#[inline(always)]
+pub fn xor3<AF: AbstractField>(x: AF, y: AF, z: AF) -> AF {
+    // The cheapest way to implement this polynomial is to simply apply xor twice.
+    // This costs 2 adds, 2 subs, 2 muls and 2 doubles.
+    xor(x, xor(y, z))
+}
+
+/// Computes the arithmetic generalization of `andnot`.
+///
+/// For boolean inputs `(!x) & y = (1 - x)y`
+#[inline(always)]
+pub fn andn<AF: AbstractField>(x: AF, y: AF) -> AF {
+    (AF::ONE - x) * y
+}
+
+/// Compute `xor` on a list of field elements using the arithmetic generalization.
+///
+/// Additionally verifies at debug time that all inputs are boolean.
+#[inline(always)]
+pub fn checked_xor<F: Field, const N: usize>(xs: [F; N]) -> F {
+    xs.into_iter().fold(F::ZERO, |acc, x| {
+        debug_assert!(x.is_zero() || x.is_one());
+        xor(acc, x)
+    })
+}
+
+/// Compute the `andnot` on a pair of field elements using the arithmetic generalization.
+///
+/// Additionally verifies at debug time that both inputs are boolean.
+#[inline(always)]
+pub fn checked_andn<F: Field>(x: F, y: F) -> F {
+    debug_assert!(x.is_zero() || x.is_one());
+    debug_assert!(y.is_zero() || y.is_one());
+    andn(x, y)
+}
+
+/// Convert a 32-bit integer into an array of 32 0 or 1 field elements.
+///
+/// The output array is in little-endian order.
+#[inline]
+pub fn u32_to_bits_le<AF: AbstractField>(val: u32) -> [AF; 32] {
+    // We do this over F::from_canonical_u32 as from_canonical_u32 can be slow
+    // like in the case of monty field.
+    array::from_fn(|i| {
+        if val & (1 << i) != 0 {
+            AF::ONE
+        } else {
+            AF::ZERO
+        }
+    })
+}
+
+/// Verify that `a = b + c + d mod 2^32`
+///
+/// We assume that a, b, c, d are all given as `2, 16` bit limbs (e.g. `a = a[0] + 2^16 a[1]`) and
+/// each `16` bit limb has been range checked to ensure it contains a value in `[0, 2^16)`.
+#[inline]
+pub fn triple_add<AB: AirBuilder>(
```

Comment:
Very cool that we can do this without committing any intermediate results!

nit: maybe `add3` to match `xor3`?